### ISSUE TITLE
RFC - possible API change to support timeouts by default

### ISSUE THREAD

**Issue by kmanc** [RsaCtfTool/RsaCtfTool#339]
Hey folks! While I was trying to add a timeout to nonRSA in https://github.com/RsaCtfTool/RsaCtfTool/pull/337, I wondered if there might be a way to wrap all of the attacks such that they run with a timeout enabled without needing to specifically add one to the attack code. It turns out, there is a way! This would allow you to remove all of the `with timeout` stuff from the attacks. It _does_ constitute a small API change which I'll explain below.

in `attacks/abstract_attack.py` , timeout and TimeoutError would be imported and the following function would be added after the existing `attack` function
```
def _attack(self, publickeys, cipher=[], progress=True):
    """Attack wrapper to implement timeouts"""
    with timeout(self.timeout):
        try:
            return self.attack(publickeys, cipher)
        except TimeoutError:
            return(None, None)
```

then in `lib/rsa_attack.py`, line 307 would be changed to
```
self.priv_key, unciphered = attack_module._attack(
```

I'll mock up a PR for this and include the cleanup that I think this can enable, but does anyone have thoughts on this change?

**Comment by daedalus** [comment]
I suggest the other way arround:

```
def attack(self, publickeys, cipher=[], progress=True):
    """Attack wrapper to implement timeouts"""
    with timeout(self.timeout):
        try:
            return self._attack(publickeys, cipher)
        except TimeoutError:
            return(None, None)
```

**Comment by kmanc** [comment]
Sorry was traveling for work, but yeah I'm happy to make that change! I'll whip something up tomorrow

**Comment by daedalus** [comment]
I understand, maybe this works better:

```
def attack_wrapper(self, publickeys, cipher=[], progress=True):
    """Attack wrapper to implement timeouts"""
    with timeout(self.timeout):
        try:
            return self.attack(publickeys, cipher)
        except TimeoutError:
            return(None, None)
```

**Comment by kmanc** [comment]
Awesome, yeah I like that. Gonna test on my Kali machine and then I'll open the PR if it goes well

**Comment by daedalus** [comment]
You are welcome, please take into account the [contributing guidelines](https://github.com/RsaCtfTool/RsaCtfTool/blob/master/CONTRIBUTING.md).

**Comment by daedalus** [comment]
Solved in PR #346
